src/lib.rs

```diff
@@ -1,4 +1,140 @@
+
+use async_read_limit::AsyncReadLimit;
+
+/// An upload limiting filter middleware for tide
+#[derive(Debug)]
+pub struct UploadLimit {
+    max_content_length: usize,
+}
+
+impl UploadLimit {
+    /// Create a new upload-limiting filter
+    #[must_use]
+    pub fn new(max_content_length: usize) -> Self {
+        Self { max_content_length }
+    }
+}
+
+/// Request body payload is larger than the configured maximum
+#[derive(Debug, thiserror::Error)]
+#[error("payload size exceeds configured maximum ({size} > {limit})")]
+pub struct Error {
+    /// The size of the payload
+    size: usize,
+
+    /// The maximum payload size
+    limit: usize,
+}
+
+impl Error {
+    pub(crate) fn new(size: usize, limit: usize) -> Self {
+        Self { size, limit }
+    }
+}
+
+#[async_trait]
+impl<State> Middleware<State> for UploadLimit
+where
+    State: Send + Sync + Clone + 'static,
+{
+    async fn handle(
+        &self,
+        mut request: tide::Request<State>,
+        next: tide::Next<'_, State>,
+    ) -> tide::Result {
+        let length = request.len();
+        check_header(self.max_content_length, length)?;
+
+        let upload_clamped = wrap_request(self.max_content_length, &mut request);
+
+        let mut response = next.run(request).await;
+
+        if upload_clamped.load(Ordering::Relaxed) {
+            response.set_status(StatusCode::PayloadTooLarge)
+        };
+
+        Ok(response)
+    }
+}
+
+/// if the length is set, and is larger than the configured maximum, then we
+/// have an 'escape hatch' without requiring any further processing.
+fn check_header(max_length: usize, length: Option<usize>) -> Result<(), tide::Error> {
+    length.map_or(Ok(()), |len| {
+        if len > max_length {
+            Err(tide::Error::new(
+                StatusCode::PayloadTooLarge,
+                Error::new(len, max_length),
+            ))
+        } else {
+            Ok(())
+        }
+    })
+}
+
+/// Wrap the request body in a byte sniffer and then reassemble the request
+fn wrap_request<State>(max_length: usize, request: &mut Request<State>) -> Arc<AtomicBool> {
+    let length = request.len();
+    let body = request.take_body();
+
+    let (sniffer, upload_clamped) = get_sniffer(max_length, body);
+
+    let sniffed_reader = tide::Body::from_reader(sniffer, length);
+
+    request.set_body(sniffed_reader);
+
+    upload_clamped
+}
+
+/// Create a new byte 'sniffer' to count bytes as they go past
+fn get_sniffer(max_length: usize, body: tide::Body) -> (impl AsyncBufRead, Arc<AtomicBool>) {
+    let upload_clamped = Arc::new(AtomicBool::new(false));
+    let upload_clamped_clone = Arc::clone(&upload_clamped);
+
+    let sniffer =
+        futures_util::io::BufReader::new(AsyncReadLimit::new(max_length, body).with_callback(
+            move |result: Result<(), async_read_limit::Error>| {
+                if result.is_err() {
+                    upload_clamped_clone.store(true, Ordering::SeqCst)
```

Comment:
This should be ordering::relaxed, not seqcst since you only need to ensure that this one value is flagged and available to other cores, you don't need to guarantee whole system ordering, especially as you load with relaxed.

Reply:
You're right, good catch